### PR TITLE
Cognitive Services 2024-06-01-preview

### PR DESCRIPTION
**Pull request by tianxin-ms**
# ARM (Control Plane) API Specification Update Pull Request 

> [!TIP]
> Overwhelmed by all this guidance? See the `Getting help` section at the bottom of this PR description.

## PR review workflow diagram

Please understand this diagram before proceeding. It explains how to get your PR approved & merged.

spec_pr_review_workflow_diagram

## Purpose of this PR

What's the purpose of this PR? Check the specific option that applies. This is **mandatory**!

  - [ ] New resource provider.
  - [x] New API version for an existing resource provider. (If API spec is not defined in TypeSpec, the PR should have been created in adherence to [OpenAPI specs PR creation guidance](https://aka.ms/azsdkdocs/createopenapispec)).
  - [ ] Update existing version for a new feature. (This is applicable only when you are revising a private preview API version.)
  - [ ] Update existing version to fix OpenAPI spec quality issues in S360.
  - [ ] Other, please clarify:
    - _edit this with your clarification_

## Due diligence checklist

To merge this PR, you **must** go through the following checklist and confirm you understood 
and followed the instructions by checking all the boxes:

- [x] I confirm this PR is modifying Azure Resource Manager (ARM) related specifications, and not data plane related specifications.
- [x] I have reviewed following [Resource Provider guidelines](https://aka.ms/rpguidelines), including
  [ARM resource provider contract](https://aka.ms/azurerpc) and
  [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md) (estimated time: 4 hours).  
  I understand this is required before I can proceed to the diagram Step 2, "ARM API changes review", for this PR.

## Additional information

<details>
<summary> Viewing API changes</summary>

For convenient view of the API changes made by this PR, refer to the URLs provided in the table 
in the `Generated ApiView` comment added to this PR. You can use ApiView to show API versions diff. 

</details>
<details>
<summary>Suppressing failures</summary>

If one or multiple validation error/warning suppression(s) is detected in your PR, please follow the 
[suppressions guide](https://aka.ms/azsdk/pr-suppressions) to get approval.

</details>

## Getting help

- First, please carefully read through this PR description, from top to bottom. Please fill out the `Purpose of this PR` and `Due diligence checklist`.
- If you don't have permissions to remove or add labels to the PR, request `write access` per [aka.ms/azsdk/access#request-access-to-rest-api-or-sdk-repositories](https://aka.ms/azsdk/access#request-access-to-rest-api-or-sdk-repositories")
- To understand what you must do next to merge this PR, see the `Next Steps to Merge` comment. It will appear within few minutes of submitting this PR and will continue to be up-to-date with current PR state.
- For guidance on fixing this PR CI check failures, see the hyperlinks provided in given failure 
  and https://aka.ms/ci-fix.
- For help with ARM review (PR workflow diagram Step 2), see https://aka.ms/azsdk/pr-arm-review.
- If the PR CI checks appear to be stuck in `queued` state, please add a comment with contents `/azp run`.
  This should result in a new comment denoting a `PR validation pipeline` has started and the checks should be updated after few minutes.
- If the help provided by the previous points is not enough, post to https://aka.ms/azsdk/support/specreview-channel and link to this PR.

### CODE REVIEW

File: specification/cognitiveservices/resource-manager/Microsoft.CognitiveServices/preview/2024-06-01-preview/cognitiveservices.json
Comment (birdemily) [inline specification/cognitiveservices/resource-manager/Microsoft.CognitiveServices/preview/2024-06-01-preview/cognitiveservices.json:6306]: If the Content Filter has multi severity levels(Low, Medium, or High).

**Review by birdemily — CHANGES_REQUESTED** [review]

**Review by birdemily — APPROVED** [review]

**Review by rkmanda — APPROVED** [review]
:shipit:

**Comment by tianxin-ms** [comment]
Hi @mikekistler , @rkmanda can we get a Signedoff on this PR? It seems the labels are abnormal, and I can edit any of them. We passed breaking change review last week.

**Comment by mentat9** [comment]
>     "RaiContentFilter": {

This API does not follow the RPC. This type needs the ARM resource envelope, with RP-specific properties in a property bag. You also need to add a singleton endpoint with GET (and PUT/DELETE if these resources are writeable). Alternatively, you could use a POST to implement the list operation, although this looks like it should be modeled as an ARM resource. You should fix this API before moving to stable. RPC ref: [https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/resource-api-reference.md#request---get-resource-collection-under-subscription](https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/resource-api-reference.md#request---get-resource-collection-under-subscription).

---
Refers to: specification/cognitiveservices/resource-manager/Microsoft.CognitiveServices/preview/2023-10-01-preview/cognitiveservices.json:5630 in 9617139. [](commit_id = 96171396bb586e4d56bb8c88a8717fd7e4f10c59, deletion_comment = False)

File: specification/cognitiveservices/resource-manager/Microsoft.CognitiveServices/preview/2023-10-01-preview/cognitiveservices.json
Comment (mentat9) [inline specification/cognitiveservices/resource-manager/Microsoft.CognitiveServices/preview/2023-10-01-preview/cognitiveservices.json:5638]: >isMultiLevelFilter [](http://example.com/codeflow?start=9&length=18)

From ARM perspective, this is moving the wrong direction. Why do you need to change from enum to boolean? Enums are generally better properties and highly recommended over booleans in ARM APIs.

File: specification/cognitiveservices/resource-manager/Microsoft.CognitiveServices/preview/2023-10-01-preview/cognitiveservices.json
Comment (mentat9) [inline specification/cognitiveservices/resource-manager/Microsoft.CognitiveServices/preview/2023-10-01-preview/cognitiveservices.json:5642]: >source [](http://example.com/codeflow?start=9&length=6)

This property (and the filterType/isMultilevelFilter property) belongs in a top-level property bag, per RPC requirements: [https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/resource-api-reference.md#request---get-resource-collection-under-subscription](https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/resource-api-reference.md#request---get-resource-collection-under-subscription).

File: specification/cognitiveservices/resource-manager/Microsoft.CognitiveServices/preview/2024-04-01-preview/cognitiveservices.json
Comment (mentat9) [inline specification/cognitiveservices/resource-manager/Microsoft.CognitiveServices/preview/2024-04-01-preview/cognitiveservices.json:6299]: >isMultiLevelFilter [](http://example.com/codeflow?start=9&length=18)

See comments on corresponding change to 2023-10-01-preview spec.

File: specification/cognitiveservices/resource-manager/Microsoft.CognitiveServices/preview/2024-04-01-preview/cognitiveservices.json
Comment (mentat9) [inline specification/cognitiveservices/resource-manager/Microsoft.CognitiveServices/preview/2024-04-01-preview/cognitiveservices.json:6303]: >source [](http://example.com/codeflow?start=9&length=6)

See comments on corresponding change to 2023-10-01-preview spec.

**Comment by mentat9** [comment]
>         "type": {

This property name is reserved by ARM at the top-level. See the RPC here: [https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/resource-api-reference.md#get-resource](https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/resource-api-reference.md#get-resource) to mean the resource type name. Your API doesn't follow the RPC, but you should avoid using reserved names (name, type, id, properties, and systemData at minimum are reserved for proxy types). RP specific properties like this one belong in the property bag (and should not duplicate ARM-reserved top-level property names).

---
Refers to: specification/cognitiveservices/resource-manager/Microsoft.CognitiveServices/preview/2024-06-01-preview/cognitiveservices.json:6150 in 9617139. [](commit_id = 96171396bb586e4d56bb8c88a8717fd7e4f10c59, deletion_comment = False)

**Comment by tianxin-ms** [comment]
Hi @mentat9, thanks for the comments, those are bug fix, to fix the swagger file to be same with our code. We've got breaking change approval in the review meeting. We will fix your comments before stable version, but could you approve this fix? Thanks!

**Comment by mentat9** [comment]
>         "isMultiLevelFilter": {

See comments on corresponding change to 2023-10-01-preview spec.

---
Refers to: specification/cognitiveservices/resource-manager/Microsoft.CognitiveServices/preview/2024-06-01-preview/cognitiveservices.json:6305 in 9617139. [](commit_id = 96171396bb586e4d56bb8c88a8717fd7e4f10c59, deletion_comment = False)

**Comment by mentat9** [comment]
>         "source": {

See comments on corresponding change to 2023-10-01-preview spec.

---
Refers to: specification/cognitiveservices/resource-manager/Microsoft.CognitiveServices/preview/2024-06-01-preview/cognitiveservices.json:6309 in 9617139. [](commit_id = 96171396bb586e4d56bb8c88a8717fd7e4f10c59, deletion_comment = False)

**Comment by tianxin-ms** [comment]
Hi @mikekistler , @rkmanda, @mentat9, we've got approval in the breaking change review meeting, but seems there is some issue of the labels, could you help with the signoff?
image

File: specification/cognitiveservices/resource-manager/Microsoft.CognitiveServices/preview/2023-10-01-preview/cognitiveservices.json
Comment (TimLovellSmith) [reply specification/cognitiveservices/resource-manager/Microsoft.CognitiveServices/preview/2023-10-01-preview/cognitiveservices.json:5638]: Yes, why change this

File: specification/cognitiveservices/resource-manager/Microsoft.CognitiveServices/preview/2023-10-01-preview/cognitiveservices.json
Comment (TimLovellSmith) [inline specification/cognitiveservices/resource-manager/Microsoft.CognitiveServices/preview/2023-10-01-preview/cognitiveservices.json:5634]: >name [](http://example.com/codeflow?start=9&length=4)

Why change it from policyName to name? PolicyName seems clear, and its a breaking change, is this reviewed?

Comment (tianxin-ms) [reply specification/cognitiveservices/resource-manager/Microsoft.CognitiveServices/preview/2023-10-01-preview/cognitiveservices.json:5634]: Thanks for the review. This parameter is not a policy name, it's a **filter name**. This breaking change is more like a bug fix, we've got the breaking change approval while the review meeting.

**Comment by TimLovellSmith** [comment]
>             "policyName"

you'll need to fix this too if you're changing it from policyName to name

---
Refers to: specification/cognitiveservices/resource-manager/Microsoft.CognitiveServices/preview/2023-10-01-preview/cognitiveservices.json:5624 in 22e9697. [](commit_id = 22e96971d2becabb6272eb154584bc2b016c9d7c, deletion_comment = False)

**Comment by TimLovellSmith** [comment]
>     "RaiContentFilter": {

This doesn't have the required 'resource' properties. The linter is probably complaining about that already.

---
Refers to: specification/cognitiveservices/resource-manager/Microsoft.CognitiveServices/preview/2024-06-01-preview/cognitiveservices.json:6398 in 22e9697. [](commit_id = 22e96971d2becabb6272eb154584bc2b016c9d7c, deletion_comment = False)

File: specification/cognitiveservices/resource-manager/Microsoft.CognitiveServices/preview/2024-06-01-preview/cognitiveservices.json
Comment (tianxin-ms) [reply specification/cognitiveservices/resource-manager/Microsoft.CognitiveServices/preview/2024-06-01-preview/cognitiveservices.json:4566]: In swagger, it use object and additionalProperties to define a dict

File: specification/cognitiveservices/resource-manager/Microsoft.CognitiveServices/preview/2024-06-01-preview/cognitiveservices.json
Comment (tianxin-ms) [reply specification/cognitiveservices/resource-manager/Microsoft.CognitiveServices/preview/2024-06-01-preview/cognitiveservices.json]: "description" is also in the definitions/kind, remove this duplicate one

File: specification/cognitiveservices/resource-manager/Microsoft.CognitiveServices/preview/2023-10-01-preview/cognitiveservices.json
Comment (rkmanda) [reply specification/cognitiveservices/resource-manager/Microsoft.CognitiveServices/preview/2023-10-01-preview/cognitiveservices.json]: [ARMBlockingComment] This needs to be fixed with the current API version itself.

**Comment by rkmanda** [comment]
>         "type": {

[ARMBlockingComment] Rename this to "RaiPolicyType"  

---
In reply to: [2183519220](https://github.com/Azure/azure-rest-api-specs/pull/29228#issuecomment-2183519220) [](http://example.com/codeflow?ancestors=2183519220)

---
Refers to: specification/cognitiveservices/resource-manager/Microsoft.CognitiveServices/preview/2024-06-01-preview/cognitiveservices.json:6150 in 9617139. [](commit_id = 96171396bb586e4d56bb8c88a8717fd7e4f10c59, deletion_comment = False)

**Comment by birdemily** [comment]
> > ```
> >     "type": {
> > ```
> 
> Rename this to "RaiPolicyType"
> 
> In reply to: [2183519220](https://github.com/Azure/azure-rest-api-specs/pull/29228#issuecomment-2183519220)
> 
> Refers to: specification/cognitiveservices/resource-manager/Microsoft.CognitiveServices/preview/2024-06-01-preview/cognitiveservices.json:6150 in [9617139](https://github.com/Azure/azure-rest-api-specs/commit/96171396bb586e4d56bb8c88a8717fd7e4f10c59). [](commit_id = [9617139](https://github.com/Azure/azure-rest-api-specs/commit/96171396bb586e4d56bb8c88a8717fd7e4f10c59), deletion_comment = False)

Hi, @rkmanda This property is not new one in this version. It is "type" since last year. Shall we need to update it in this version?
image

**Comment by madhura-np** [comment]
>         "source": {

This was marked [ARMBlockingComment] on the 2023-10-01-preview json. i see the file has been removed, but the same comment on top level properties applies here as well. 

---
In reply to: [2183525798](https://github.com/Azure/azure-rest-api-specs/pull/29228#issuecomment-2183525798) [](http://example.com/codeflow?ancestors=2183525798)

---
Refers to: specification/cognitiveservices/resource-manager/Microsoft.CognitiveServices/preview/2024-06-01-preview/cognitiveservices.json:6309 in 9617139. [](commit_id = 96171396bb586e4d56bb8c88a8717fd7e4f10c59, deletion_comment = False)

**Comment by madhura-np** [comment]
>     "RaiContentFilter": {

This comment applies to 2024-04-01-preview as well.

---
In reply to: [2183505277](https://github.com/Azure/azure-rest-api-specs/pull/29228#issuecomment-2183505277) [](http://example.com/codeflow?ancestors=2183505277)

---
Refers to: specification/cognitiveservices/resource-manager/Microsoft.CognitiveServices/preview/2023-10-01-preview/cognitiveservices.json:5630 in 9617139. [](commit_id = 96171396bb586e4d56bb8c88a8717fd7e4f10c59, deletion_comment = False)

File: specification/cognitiveservices/resource-manager/Microsoft.CognitiveServices/preview/2023-10-01-preview/cognitiveservices.json
Comment (birdemily) [reply specification/cognitiveservices/resource-manager/Microsoft.CognitiveServices/preview/2023-10-01-preview/cognitiveservices.json]: We will take the suggests from https://github.com/Azure/azure-rest-api-specs/pull/29228#issuecomment-2183505277. To use "post" action instead of "get". We won't treat "RaiContentFilter"  as a ARM resource. In this case, shall we need to move these properties?

**Comment by birdemily** [comment]
> > ```
> >     "source": {
> > ```
> 
> This was marked [ARMBlockingComment] on the 2023-10-01-preview json. i see the file has been removed, but the same comment on top level properties applies here as well.
> 
> In reply to: [2183525798](https://github.com/Azure/azure-rest-api-specs/pull/29228#issuecomment-2183525798)
> 
> Refers to: specification/cognitiveservices/resource-manager/Microsoft.CognitiveServices/preview/2024-06-01-preview/cognitiveservices.json:6309 in [9617139](https://github.com/Azure/azure-rest-api-specs/commit/96171396bb586e4d56bb8c88a8717fd7e4f10c59). [](commit_id = [9617139](https://github.com/Azure/azure-rest-api-specs/commit/96171396bb586e4d56bb8c88a8717fd7e4f10c59), deletion_comment = False)

We will take the suggests from https://github.com/Azure/azure-rest-api-specs/pull/29228#issuecomment-2183505277. To use "post" action instead of "get". We won't treat "RaiContentFilter"  as a ARM resource. In this case, shall we need to move these properties?

File: specification/cognitiveservices/resource-manager/Microsoft.CognitiveServices/preview/2024-06-01-preview/cognitiveservices.json
Comment (birdemily) [inline specification/cognitiveservices/resource-manager/Microsoft.CognitiveServices/preview/2024-06-01-preview/cognitiveservices.json:2824]: get content filter by name

File: specification/cognitiveservices/resource-manager/Microsoft.CognitiveServices/preview/2024-06-01-preview/examples/ListRaiContentFilters.json
Comment (birdemily) [inline specification/cognitiveservices/resource-manager/Microsoft.CognitiveServices/preview/2024-06-01-preview/examples/ListRaiContentFilters.json:104]: Delete this one

**Comment by raosuhas** [comment]
>         "type": {

Still active 

---
In reply to: [2218399517](https://github.com/Azure/azure-rest-api-specs/pull/29228#issuecomment-2218399517) [](http://example.com/codeflow?ancestors=2218399517,2183519220)

---
Refers to: specification/cognitiveservices/resource-manager/Microsoft.CognitiveServices/preview/2024-06-01-preview/cognitiveservices.json:6150 in 9617139. [](commit_id = 96171396bb586e4d56bb8c88a8717fd7e4f10c59, deletion_comment = False)

**Comment by raosuhas** [comment]
You still have a list of warnings in the swagger lint : https://github.com/Azure/azure-rest-api-specs/pull/29228/checks?check_run_id=27596033064 

Please fix those as well. (You can ignore the one for "[AvoidNestedProperties]")

**Comment by tianxin-ms** [comment]
Hi @raosuhas , for the "type", we left the comment earlier, this property is not new in the current version. It is "type" since last year. If we rename it, it will cause breaking change, should we still rename it? 
I fixed some lint error, for others, similar reason, it will cause other issue if we use latest common type. 

This part is more like a fix/update on an existed API, could you help to check it again?

**Pull request merged**
